src/validationinterface.cpp

```diff
@@ -128,52 +132,69 @@ void SyncWithValidationInterfaceQueue() {
     promise.get_future().wait();
 }
 
-void CMainSignals::MempoolEntryRemoved(CTransactionRef ptx, MemPoolRemovalReason reason) {
-    if (reason != MemPoolRemovalReason::BLOCK && reason != MemPoolRemovalReason::CONFLICT) {
-        m_internals->m_schedulerClient.AddToProcessQueue([ptx, this] {
-            m_internals->TransactionRemovedFromMempool(ptx);
-        });
-    }
-}
-
 void CMainSignals::UpdatedBlockTip(const CBlockIndex *pindexNew, const CBlockIndex *pindexFork, bool fInitialDownload) {
     // Dependencies exist that require UpdatedBlockTip events to be delivered in the order in which
     // the chain actually updates. One way to ensure this is for the caller to invoke this signal
     // in the same critical section where the chain is updated
 
-    m_internals->m_schedulerClient.AddToProcessQueue([pindexNew, pindexFork, fInitialDownload, this] {
+    auto func = __func__;
+    m_internals->m_schedulerClient.AddToProcessQueue([func, pindexNew, pindexFork, fInitialDownload, this] {
```

Comment:
For all of these async validationinterface callbacks, I think it'd be useful to log both when the event is enqueued and dequeued, with some kind of counter to correlate the two logs. Thoughts?

Reply:
The scheduler client guarantees these callbacks are executed serially in the order in which they were enqueued. Unless there is a need to correlate the enqueuing with the execution of `BlockChecked` and `NewPowValidBlock`, I'm hesitant to add more complexity to the code.

Is there a specific case where you think the additional logging would be useful? If so, would correlation using data from the log message be sufficient rather than adding a counter?

Reply:
I think this: https://github.com/bitcoin/bitcoin/issues/12978#issuecomment-381265760 would have been easier to spot with logging when the callbacks were enqueued, and it would have been easier to correlate those enqueue events with the thread that was running.

You're probably right that there's no need for a counter and that the data from the log message is enough to correlate the enqueue/dequeue events.

Reply:
Done. In order to avoid duplicating the logging code, I added some private template methods declared in the header and defined in the implementation file. The linker seems ok with this though let me know if there is a better way. Defining the helpers in the header would have required moving `MainSignalsInstance` into the header.

Reply:
FYI, I had to make some minor changes to placate the linter. It was giving the following failures:

```
src/validationinterface.cpp: Expected 0 argument(s) after format string but found 1 argument(s): LogPrint(BCLog::VALIDATIONINTERFACE, fmt, args...)
^---- failure generated from test/lint/lint-format-strings.sh
All calls to LogPrintf() and LogPrint() should be terminated with \n
src/validationinterface.cpp:    LogPrint(BCLog::VALIDATIONINTERFACE, fmt, args...);
^---- failure generated from test/lint/lint-logs.sh
```

https://travis-ci.org/bitcoin/bitcoin/jobs/599302990

Reply:
> In order to avoid duplicating the logging code, I added some private template methods declared in the header and defined in the implementation file. The linker seems ok with this(...)

This could lead to portability issues, see e.g. https://stackoverflow.com/questions/495021/why-can-templates-only-be-implemented-in-the-header-file. I tried something similar recently in another PR, and the linker in one of the Travis runs did complain, so I am also interested to know if there are better solutions than moving stuff into the header.

Reply:
Yeah, that could very well be the case. That said, if we want to avoid argument evaluation when the logging category is not enabled, I would need to turn these into macros. See discussion starting in https://github.com/bitcoin/bitcoin/pull/16688#issuecomment-543388763. Then I could remove those template method declarations from the header.

Reply:
This does indeed seem to have caused a build failure: https://travis-ci.org/bitcoin/bitcoin/jobs/600461225

Reply:
Indeed. It seems variadic arguments within lambda captures are not supported by that compilation environment. At least that's what I've gleaned from a similar issue (https://github.com/Ultimaker/CuraEngine/issues/426#issuecomment-269451925). However, this shouldn't be a problem if I change it into a macro.

Reply:
Turns out changing `EnqueueAndLogEvent` into a macro has some implications that are making this infeasible. Consider:

```c++
#define ENQUEUE_AND_LOG_EVENT(event, fmt, ...)                                             \
    do {                                                                                   \
        LOG_EVENT("Enqueuing " fmt, __VA_ARGS__);                                          \
        m_internals->m_schedulerClient.AddToProcessQueue([event, fmt, __VA_ARGS__, this] { \
            LOG_EVENT(fmt, __VA_ARGS__);                                                   \
            event();                                                                       \
        });                                                                                \
    } while (0)

#define LOG_EVENT(fmt, ...) \
    LogPrint(BCLog::VALIDATIONINTERFACE, fmt "\n", __VA_ARGS__);
```

The problem is that the capture list is expecting variables but the call sites are providing expressions. So unfortunately this won't compile. Presumably, if this did work, the expressions in the lambda would be evaluated thus defeating the purpose.

Unless there is a creative solution that I'm unaware of, the only alternative I can think of is to forgo the macro and duplicate the `LogPrint` for the enqueuing in each method. Thoughts?

Reply:
Actually, it looks like using a default capture instead of __VA_ARGS__ will do what I need!

Reply:
Implemented as a macro in 6ab424c.